02-peer-protocol.md

```diff
@@ -50,10 +67,451 @@ pubkey corresponding to the funding output nothing prevents duplicative channel
+        |       |<-(7)- tx_complete ----|       |
+        +-------+                       +-------+
+
+#### `initiator` and `contributor` contributions
+
+In this example, both A the `initiator` and B the `contributor` will add
+funds to the transaction. A contributes 2 inputs and an output that they
+later remove; B contributes 1 input and an output, but waits until A
+adds an input with commitment proof data before contributing. In this
+example, A's second transmitted input will contain the commitment data.
+
+        +-------+                       +-------+
+        |       |--(1)- tx_add_input -->|       |
+        |       |<-(2)- tx_complete ----|       |
+        |       |--(3)- tx_add_output ->|       |
+        |       |<-(4)- tx_complete ----|       |
+        |       |--(5*) tx_add_input -->|       |
+        |   A   |<-(6)- tx_add_input ---|   B   |
+        |       |<-(7)- tx_add_output --|       |
+        |       |--(8)- tx_rm_output -->|       |
+        |       |<-(9)- tx_complete ----|       |
+        |       |--(10) tx_complete --->|       |
+        +-------+                       +-------+
+
+\* This `tx_add_input` includes a commitment proof.
+
+
+### The `tx_add_input` Message
+
+This message contains information for inputs to a funding transaction.
+
+1. type: 66 (`tx_add_input`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`u16`:`serial_id`]
+    * [`u16`:`prevtx_len`]
+    * [`prevtx_len*byte`:`prevtx`]
+    * [`u32`:`prevtx_vout`]
+    * [`u16`:`max_witness_len`]
+    * [`u16`:`redeemscript_len`]
+    * [`redeemscript_len*byte`:`script`]
+    * [`tx_add_input_tlvs`:`tlvs`]
+
+1. tlvs: `tx_add_input_tlvs`
+2. types:
+   1. type: 2 (`podle_proof`)
+   2. data:
+       * [`point`:`p`]
+       * [`point`:`p2`]
+       * [`sha256`:`e`]
+       * [`32*byte`:`sig`]
+
+#### Requirements
+
+The sending node:
+  - MUST add all sent inputs to the transaction
+  - MUST use a unique `serial_id` for each input
+  - MUST NOT re-transmit inputs it has already received from the peer
+  - MUST send a non-malleable (segwit) UTXO.
+  - if is the `initiator`:
+    - MUST send even `serial_id`s
+  - if is the `contributor`:
+    - MAY omit this message
+    - MUST send odd `serial_id`s
+
+The receiving node:
+  - MUST add all received inputs to the funding transaction
+  - MUST fail the transaction collaboration if:
+    - it receives a duplicate input to one it sent previously
+    - it receives an input that would create a malleable transaction id (e.g. pre-Segwit)
+    - it receives an unconfirmed input
+    - it receives a duplicate `serial_id`
+    - it receives a `serial_id` from the peer with the incorrect parity
+
+#### Rationale
+Each node must have a complete set of the transaction inputs
+to derive the funding transaction and subsequent commitment signatures.
+
+`serial_id` is a randomly chosen number which uniquely identifies this input.
+Inputs in the final transaction will be sorted by `serial_id`.
```

Comment:
Are there any privacy concerns around letting users pick these numbers as opposed to say using a hash of information that is less in the users' control? I figured I'd ask but I think the answer is no since a malicious party can always just leak this information to the public if they wanted

Reply:
The motivation for allowing for selection of serial_ids is that it allows a side to (potentially, there's definitely edge cases where it's not possible) spend a SIGHASH_SINGLE sig inputs